### PR TITLE
Expand table-like input options for pygmt.surface

### LINKED ISSUE #1443: Expand table-like input options for pygmt.surface
**Description of the desired feature**

Currently, the `data` parameter for [pygmt.surface](https://www.pygmt.org/dev/api/generated/pygmt.surface.html) function accepts a data file name or a 2d numpy array with the tabular data. It would be helpful to expand the list of accepted input options to include pandas.DataFrame, xarray.Dataset, or geopandas.GeoDataFrame similar to [blockmedian](https://www.pygmt.org/dev/api/generated/pygmt.blockmedian.html) or [plot](https://www.pygmt.org/dev/api/generated/pygmt.Figure.plot.html).

To address, this issue you will need to change this block of code:
https://github.com/GenericMappingTools/pygmt/blob/cd822ca2c01f98a1496c63cc0bff8cc6661cb3bb/pygmt/src/surface.py#L87-L94
to:
```
            # Choose how data will be passed into the module
            file_context = lib.virtualfile_from_data(
                check_kind="vector", data=data, x=x, y=y, z=z
            )
```

update the `data` parameter description in the docstring to:
```
    data : str or {table-like}
        Pass in (x, y, z) or (longitude, latitude, elevation) values by
        providing a file name to an ASCII data table, a 2D
        {table-classes}.
```
and remove `dummy_context` from the import statements in pygmt/src/surface.py.

**Are you willing to help implement and maintain this feature?** I would be glad to help someone implement this feature request.

### PR DESCRIPTION
**Pull request by daroari**
**Description of proposed changes**

This PR expands the table-like input options for pygmt.surface

Closes #1443 

**Reminders**

- [x] Run `make format` and `make check` to make sure the code follows the style guide.
- [ ] Add tests for new features or tests that would have caught the bug that you're fixing.
- [ ] Add new public functions/methods/classes to `doc/api/index.rst`.
- [ ] Write detailed docstrings for all functions/methods.
- [ ] If adding new functionality, add an example to docstrings or tutorials.

**Slash Commands**

You can write slash commands (`/command`) in the first line of a comment to perform
specific operations. Supported slash commands are:

- `/format`: automatically format and lint the code
- `/test-gmt-dev`: run full tests on the latest GMT development version

### CODE REVIEW

File: pygmt/src/surface.py [hunk 1]
```diff
@@ -84,14 +84,10 @@ def surface(x=None, y=None, z=None, data=None, **kwargs):
...
-                file_context = dummy_context(data)
-            elif kind == "matrix":
-                file_context = lib.virtualfile_from_matrix(data)
-            elif kind == "vectors":
-                file_context = lib.virtualfile_from_vectors(x, y, z)
-            else:
-                raise GMTInvalidInput("Unrecognized data type: {}".format(type(data)))
+            # Choose how data will be passed into the module
+            file_context = lib.virtualfile_from_data(
+                    kind = "vector", data=data, x=x, y=y, z=z
```
Comment (meghanrjones) [inline pygmt/src/surface.py:89]:
```suggestion
                    check_kind="vector",  data=data, x=x, y=y, z=z
```

**Review by meghanrjones — COMMENTED** [review]
Great work!

Can you please remove this line from /pygmt/src/surface.py in your branch? This function is no longer used in the module, so it does not need to be imported:

https://github.com/GenericMappingTools/pygmt/blob/7a94b683e74b4bdc2cde3a0b0ccf124cd846ac8c/pygmt/src/surface.py#L12

**Comment by seisman** [comment]
Ping @daroari to address the comment https://github.com/GenericMappingTools/pygmt/pull/1455#pullrequestreview-735597358

**Comment by meghanrjones** [comment]
Hi @daroari, do you think you will have time to finish this pull request? It would be great to get it added to PyGMT 😄

If not, we can either close the PR or I could push the remaining changes needed to this branch, whichever you prefer.

**Comment by daroari** [comment]
Hi @meghanrjones. Sorry for the delayed response. Would you please push the remaining changes? I was going to do it, but I am not able to do it right now. Sorry for the inconvenience :(

**Review by weiji14 — APPROVED** [review]
Thumbs up from me. Technically, it would be nice to have a unit test to make sure this works (e.g. using [`pytest.mark.parametrize`](https://docs.pytest.org/en/6.2.x/parametrize.html) like for blockm* in #1523), but ok to do it another time too :slightly_smiling_face:

**Review by meghanrjones — APPROVED** [review]
Thanks @daroari for starting the work and @seisman for finishing it up!

**Pull request merged**